library/src/main/java/com/datatorrent/lib/async/AbstractAsyncProcessor.java

```diff
@@ -0,0 +1,341 @@
+ * </ol>
+ *
+ *
+ * <b>Use case examples:</b>
+ * <ul>
+ *   <li>Parallel reads from external datastore/database system.</li>
+ *   <li>Operations which are long running tasks per tuple but DAG i/o should not be blocked.</li>
+ * </ul>
+ *
+ * @param <QUEUETUPLE> input type of tuple
+ * @param <RESULTTUPLE> Result of async processing of tuple.
+ */
+public abstract class AbstractAsyncProcessor<QUEUETUPLE, RESULTTUPLE> extends BaseOperator implements IdleTimeHandler
+{
+  private static final Logger logger = LoggerFactory.getLogger(AbstractAsyncProcessor.class);
+
+  /**
+   * This is an executor service which holds thread pool and enqueues task for processing.
+   */
+  private transient ExecutorService executor;
+
+  /**
+   * Number of worker threads for processing
+   * Minimum value is 1 and max allowed number of threads are 32.
+   *
+   * Default number of threads is 32.
+   */
+  @Min(1)
+  @Max(32)
+  private int numProcessors = 32;
+
+  /**
+   * Tells whether the <i>handleProcessedTuple</i> should be called for tuples in the same order in which
+   * they get enqueued using <i>enqueueTupleForProcessing</i>.
+   */
+  private boolean maintainTupleOrder = false;
+
+  /**
+   * Timeout in milliseconds after which processing of tuple may be terminated.
+   */
+  private long processTimeoutMs = 0;
+
+  /**
+   * The queue holds following types of tuples:
+   * <ul>
+   *   <li>Yet to be processed</li>
+   *   <li>In Process</li>
+   *   <li>Completed processing but yet to be notified for process completion</li>
+   * </ul>
+   */
+  private Queue<WorkItem> waitingTuples = Lists.newLinkedList();
+
+  /**
+   * Setup method of this operator will initialize various types of executors possible.
+   *
+   * @param context OperatorContext
+   */
+  @Override public void setup(Context.OperatorContext context)
+  {
+    this.executor = Executors.newFixedThreadPool(numProcessors);
+    replayTuplesIfRequired();
+  }
+
+  /**
+   * This method is called from setup of the operator. If <i>waitingTuples</i> contains anything, the method enqueues
+   * all those tuples for processing to worker threads.
+   *
+   * Dependening on condition the method will behave as follows:
+   * <ul>
+   *   <li>During fresh start of operator, this method will not have any effect as there are no tuples in
+   *       <i>waitingTuples</i>.</li>
+   *   <li>During restore of operator to previous checkpointed state, if there are any tuples already present
+   *       in <i>waitingTuples</i> that means they're yet to be notified, hence enqueued again.</li>
+   * </ul>
+   */
+  private void replayTuplesIfRequired()
+  {
+    // Enqueue everything that is still left.
+    // Its left because its either not completed processing OR its not emitted because of ordering restrictions.
+    for (WorkItem item : waitingTuples) {
+      item.taskHandle = executor.submit(new Processor(item));
+      item.processState = State.SUBMITTED;
```

Comment:
calling executor.submit after variables are set. This should take care of overrides.